lib/Dialect/Triton/Transforms/LoopPeeling.cpp

```diff
@@ -0,0 +1,74 @@
+#include "triton/Dialect/Triton/Transforms/LoopPeeling.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Pass/Pass.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace triton {
+
+void peelLoopEpilogue(
+    scf::ForOp forOp,
+    function_ref<Operation *(RewriterBase &, Operation *, bool)>
+        processPeeledOp) {
+  SmallVector<Operation *> loopBodyOps;
+  IRRewriter rewriter(forOp);
+  Location loc = forOp.getLoc();
+  Type type = forOp.getStep().getType();
+
+  // Fetch loop bounds and step
+  Value lowerBound = forOp.getLowerBound();
+  Value upperBound = forOp.getUpperBound();
+  Value step = forOp.getStep();
+  Value newUpperBound = rewriter.create<arith::SubIOp>(loc, upperBound, step);
+  forOp.getUpperBoundMutable().assign(newUpperBound);
+
+  rewriter.setInsertionPointAfter(forOp);
+
+  // Last iter induction variable value
+  // lastIV = lb + floor( (ub – lb – 1) / s ) * s
+  Value range = rewriter.create<arith::SubIOp>(loc, upperBound, lowerBound);
```

Comment:
nit: can we refactor this with https://github.com/triton-lang/triton/blob/75fe1135f4cbc7fefff02b6cf497d340fbc7d9cc/lib/Dialect/TritonGPU/Transforms/WarpSpecialization/LoadMMASpecialization.cpp#L458